<a href="https://colab.research.google.com/github/chelseyyenmacaspac/Exercise2/blob/main/Exercise2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Exercise 2: Code Generation with ReACT Prompting

## Goal
Use a ReACT-style process to plan, generate, run, observe, and improve Python code.

## Tool Used
Google Colab and Python

## ReACT Prompt

You are a Python coding assistant. Create a Python program that calculates the average of a list of student test scores.

Follow this ReACT-style process:

1. PLAN: Explain briefly how you will solve the problem.
2. ACT: Generate simple Python code to calculate the average.
3. RUN: Execute the code using these scores: 85, 90, 78, 92, 88.
4. OBSERVE: Check whether the output is correct and identify any errors.
5. FIX: If there is an error, correct the code and run it again.

Requirements:
- Use only basic Python.
- The program must accept a list of numbers.
- Handle an empty list without causing an error.
- Display the average rounded to two decimal places.
- Keep the code simple and easy to understand.

## Step 1: PLAN

The program will store the student test scores in a list. It will check whether the list is empty to prevent a division-by-zero error. If scores are available, it will calculate the average by adding the scores and dividing by the number of scores. The final average will be displayed with two decimal places.

In [1]:
def calculate_average(scores):
    if len(scores) == 0:
        return "No scores available."

    average = sum(scores) / len(scores)
    return round(average, 2)


scores = [85, 90, 78, 92, 88]

result = calculate_average(scores)

print("Student Scores:", scores)
print("Average Score:", result)

Student Scores: [85, 90, 78, 92, 88]
Average Score: 86.6


## Step 3: RUN and OBSERVE

The code ran successfully using the test scores 85, 90, 78, 92, and 88.

The calculated average was 86.6. The program worked correctly, but I noticed that the output did not display exactly two decimal places as required. It showed 86.6 instead of 86.60.

I will improve the formatting in the next step.

## Step 4: FIX

I updated the code so the average always displays exactly two decimal places. I also tested an empty list to make sure the program handles it without causing an error.

In [2]:
def calculate_average(scores):
    if len(scores) == 0:
        return None

    return sum(scores) / len(scores)


# Test 1: Normal list
scores = [85, 90, 78, 92, 88]
result = calculate_average(scores)

if result is not None:
    print("Student Scores:", scores)
    print(f"Average Score: {result:.2f}")
else:
    print("No scores available.")


# Test 2: Empty list
empty_scores = []
empty_result = calculate_average(empty_scores)

print("\nEmpty List Test:")

if empty_result is not None:
    print(f"Average Score: {empty_result:.2f}")
else:
    print("No scores available.")

Student Scores: [85, 90, 78, 92, 88]
Average Score: 86.60

Empty List Test:
No scores available.


## Step 5: Final Observation

After fixing the formatting, the program successfully displayed the average as 86.60. I also tested the program with an empty list, and it returned "No scores available" instead of causing an error.

The final code successfully completed the ReACT process: Plan → Act → Run → Observe → Fix.